src/Storage/Device/Local.php

```diff
@@ -119,26 +111,120 @@ public function upload(string $source, string $path, int $chunk = 1, int $chunks
-                }
-                \unlink($part);
+            $this->joinChunks($path, $chunks);
+
+            return $chunksReceived;
+        }
+
+        return $chunksReceived;
+    }
+
+    /**
+     * Upload Data.
+     *
+     * Upload file contents to desired destination in the selected disk.
+     * return number of chunks uploaded or 0 if it fails.
+     *
+     * @param  string  $source
+     * @param  string  $path
+     * @param  string  $contentType
+     * @param int chunk
+     * @param int chunks
+     * @param  array  $metadata
+     * @return int
+     *
+     * @throws \Exception
+     */
+    public function uploadData(string $data, string $path, string $contentType, int $chunk = 1, int $chunks = 1, array &$metadata = []): int
+    {
+        $this->createDirectory(\dirname($path));
+
+        if ($chunks === 1) {
+            if (! \file_put_contents($path, $data)) {
+                throw new Exception('Can\'t write file '.$path);
             }
-            \unlink($tmp);
+
+            return $chunks;
+        }
+        $tmp = \dirname($path).DIRECTORY_SEPARATOR.'tmp_'.\basename($path).DIRECTORY_SEPARATOR.\basename($path).'_chunks.log';
+
+        $this->createDirectory(\dirname($tmp));
+        if (! file_put_contents($tmp, "$chunk\n", FILE_APPEND)) {
+            throw new Exception('Can\'t write chunk log '.$tmp);
+        }
+
+        $chunkLogs = file($tmp);
+        if (! $chunkLogs) {
+            throw new Exception('Unable to read chunk log '.$tmp);
+        }
+
+        $chunksReceived = count(file($tmp));
+
+        if (! \file_put_contents(dirname($tmp).DIRECTORY_SEPARATOR.pathinfo($path, PATHINFO_FILENAME).'.part.'.$chunk, $data)) {
+            throw new Exception('Failed to write chunk '.$chunk);
+        }
+
+        if ($chunks === $chunksReceived) {
+            $this->joinChunks($path, $chunks);
 
             return $chunksReceived;
         }
 
         return $chunksReceived;
     }
 
+    private function joinChunks(string $path, int $chunks)
+    {
+        $tmp = \dirname($path).DIRECTORY_SEPARATOR.'tmp_'.\basename($path).DIRECTORY_SEPARATOR.\basename($path).'_chunks.log';
+        for ($i = 1; $i <= $chunks; $i++) {
+            $part = dirname($tmp).DIRECTORY_SEPARATOR.pathinfo($path, PATHINFO_FILENAME).'.part.'.$i;
+            $data = file_get_contents($part);
+            if (! $data) {
+                throw new Exception('Failed to read chunk '.$part);
+            }
+
+            if (! file_put_contents($path, $data, FILE_APPEND)) {
+                throw new Exception('Failed to append chunk '.$part);
+            }
+            \unlink($part);
+        }
+        \unlink($tmp);
+    }
+
+    /**
+     * Transfer
+     *
+     * @param  string  $path
+     * @param  string  $destination
+     * @param  Device  $device
+     * @return string
+     */
+    public function transfer(string $path, string $destination, Device $device): bool
+    {
```

Comment:
I think we should not pass the fullpath but use the device root?
$destination = $device->getRoot() . '/' . $destination;

Reply:
The problem is we don't do it yet for upload, write, read, so changing it here only would be confusing.

Reply:
if we change everywhere to add `root` automatically to the provided path, we will have to adjust a lot in current Appwrite implementation

Reply:
👍